drivers/char/hw_random/bcm2835-rng.c

```diff
@@ -65,23 +69,33 @@ static int bcm2835_rng_read(struct hwrng *rng, void *buf, size_t max,
 			       bool wait)
 {
 	struct bcm2835_rng_priv *priv = to_rng_priv(rng);
+	unsigned long idle_endtime = jiffies + HZ; /* Wait up to 1 second */
 	u32 max_words = max / sizeof(u32);
-	u32 num_words, count;
+	u32 num_words, count = 0;
 
-	while ((rng_readl(priv, RNG_STATUS) >> 24) == 0) {
-		if (!wait)
-			return 0;
-		hwrng_msleep(rng, 1000);
-	}
+	while ((count < max_words) && time_before(jiffies, idle_endtime)) {
+		while ((rng_readl(priv, RNG_STATUS) >> 24) == 0) {
+			if (!wait) {
+				max_words = count;
+				break;
+			}
+
+			/* Can wait, give others chance to run */
+			num_words = min(max_words - count, (u32)RNG_FIFO_WORDS);
+			usleep_range((u32)RNG_US_PER_WORD, num_words * RNG_US_PER_WORD);
+		}
 
-	num_words = rng_readl(priv, RNG_STATUS) >> 24;
-	if (num_words > max_words)
-		num_words = max_words;
+		num_words = min(rng_readl(priv, RNG_STATUS) >> 24,
+				max_words - count);
 
-	for (count = 0; count < num_words; count++)
-		((u32 *)buf)[count] = rng_readl(priv, RNG_DATA);
+		for (; num_words; num_words--, count++)
+			((u32 *)buf)[count] = rng_readl(priv, RNG_DATA);
+
+		/* Reset the IDLE timeout */
+		idle_endtime = jiffies + HZ;
```

Comment:
Doesn't make this the timeout test in the outer loop useless?

Reply:
As the commit message says, the timeout protects against no data arriving for a second, not against whether the whole read takes more than a second.